core/corerepo/gc.go

```diff
@@ -66,3 +71,110 @@ func GarbageCollectAsync(n *core.IpfsNode, ctx context.Context) (<-chan *KeyRemo
 	}()
 	return output, nil
 }
+
+func PeriodicGC(ctx context.Context, node *core.IpfsNode) error {
+	r := node.Repo
+	cfg := r.Config()
+	if cfg.Datastore.GCPeriod == "" {
+		r.SetConfigKey("Datastore.GCPeriod", "1h")
+		cfg.Datastore.GCPeriod = "1h"
+	}
+
+	period, err := time.ParseDuration(cfg.Datastore.GCPeriod)
+	if err != nil {
+		return err
+	}
+	if int64(period) == 0 {
+		// if duration is 0, it means GC is disabled.
+		return nil
+	}
+
+	storageMax, storageGC, slackGB, err := getGCParam(node.Repo)
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(period):
+			// the private func maybeGC doesn't compute storageMax, storageGC, slackGC so that they are not re-computed for every cycle
+			if err := maybeGC(ctx, node, storageMax, storageGC, slackGB, 0); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func ConditionalGC(ctx context.Context, node *core.IpfsNode, offset uint64) error {
+	storageMax, storageGC, slackGB, err := getGCParam(node.Repo)
+	if err != nil {
+		return err
+	}
+	return maybeGC(ctx, node, storageMax, storageGC, slackGB, offset)
+}
+
+func getGCParam(r repo.Repo) (storageMax uint64, storageGC uint64, slackGB uint64, err error) {
+	cfg := r.Config()
+	// check if cfg has these fields initialized
+	// TODO: there should be a general check for all of the cfg fields
+	// maybe distinguish between user config file and default struct?
+	if cfg.Datastore.StorageMax == "" {
+		r.SetConfigKey("Datastore.StorageMax", "10GB")
+		cfg.Datastore.StorageMax = "10GB"
+	}
+	if cfg.Datastore.StorageGCWatermark == 0 {
+		r.SetConfigKey("Datastore.StorageGCWatermark", 90)
+		cfg.Datastore.StorageGCWatermark = 90
+	}
+
+	storageMax, err = humanize.ParseBytes(cfg.Datastore.StorageMax)
+	if err != nil {
+		return
+	}
+	storageGC = storageMax * uint64(cfg.Datastore.StorageGCWatermark) / 100
+
+	// calculate the slack space between StorageMax and StorageGCWatermark
+	// used to limit GC duration
+	slackGB = (storageMax - storageGC) / 10e9
+	if slackGB < 1 {
+		slackGB = 1
+	}
```

Comment:
then again, slack is only used for bounding GC time, so this case doesnt matter for smaller memories.